src/editor/classes/UiDriver.js

```diff
@@ -1,31 +1,160 @@
+    
+
+    // Helper functions for hooks
+    this.detectIndentationMode = function(editor) {
+        var len = editor.lineCount();
+        var regexIndented = /^([ ]{2,}|[\t]+)[^ \t]+?/g; // Is not blank, and is indented with tab or space
+
+        for (var i = 0; i < len && i < 100; i++) {
+            var line = editor.getLine(i);
+            var matches = regexIndented.exec(line);
+            if (matches !== null) {
+                if (line[0] === "\t") { // Is a tab
+                    return [1, 0];
+                } else { // Is a space
+                    var size = matches[1].length;
+                    if (size === 2 || size === 4 || size === 8) {
+                        return [0, size];
+                    } else {
+                        return undefined;
+                    }
+                }
+            }
+        }
+        return undefined;
+
+    }
+
+    this.registerEventHandler = function(msg, handler) {
+        if (handlers[msg] === undefined)
+            handlers[msg] = [];
+        handlers[msg].push(handler);
+    }
+
+    this.setReturnData = function(data) {
+        this.proxy.result = data;
+    }
+    
+    this.messageReceived = function(msg, data) {
+        if (!usingQtWebChannel()) {
+            console.error("Not using QtWebChannel: "+ msg);
+        }
+
+        // Only one of the handlers (the last that gets
+        // called) can return a value. So, to each handler
+        // we provide the previous handler's return value.
+        var prevReturn = undefined;
+        if (handlers[msg] !== undefined) {
+            handlers[msg].forEach(function(handler) {
+                prevReturn = handler(msg, data, prevReturn);
+            });
+        }
+        
+        if(prevReturn !== undefined) {
+            this.setReturnData(prevReturn);
+        }
+        return prevReturn;
+    }
+
+    // BEGIN HOOKS
+    // These functions hook into the CodeMirror interface and allow us to
+    // perform asynchronous data transfer between C++ and Javascript.
+
+    this.onOptionChange = function(editor) {
+        var indentMode = [editor.options.indentWithTabs, editor.options.indentUnit];
+        this.proxy.setValue("indentMode", indentMode);
+    }
+
+    // Hook for when the editor's language is changed.
+    this.onLanguageChange = function(editor) {
+        var langId = Languages.currentLanguage(editor);
+        var langData =  {id: langId, lang: Languages.languages[langId]}; 
+        this.proxy.setValue("language", langData);
+        this.proxy.sendEditorEvent("J_EVT_CURRENT_LANGUAGE_CHANGED", langData);
+    }
+
+    // Hook for when we load a file/change content
+    this.onSetValue = function(editor) {
+        var detectedIndent = this.detectIndentationMode(editor);
+        if (detectedIndent !== undefined) {
+            this.proxy.setValue("detectedIndent", detectedIndent);
+        }
+        editor.clearHistory();
+        this.proxy.sendEditorEvent("J_EVT_FILE_LOADED", 0);
```

Comment:
"J_EVT_FILE_LOADED" is misleading: I suggest "J_EVT_DOCUMENT_LOADED".
Same thing for the associated events and slots.